drivers: serial: ns16550: Fix TX IRQ not triggered when FIFO is empty

jackylee-go opened the pull request:
### Summary

This PR fixes an issue in the `uart_ns16550` driver where calling
`uart_ns16550_irq_tx_enable()` while the TX FIFO is already empty
fails to generate a new interrupt, potentially stalling transmission.

### Details

- The issue occurs when the TX FIFO is empty and no interrupt is pending.
- The workaround ensures that an ISR callback would be executed when the 
   FIFO is empty at the time of enabling TX IRQ.
- This helps maintain consistent behavior in interrupt-driven UART TX.

### Testing

- Verified on Andes AE350
- Confirmed that TX interrupt is now reliably triggered even when FIFO is empty

### Notes

Let me know if this workaround should be made optional via Kconfig or if
there are concerns about side effects on other platforms.

---

Signed-off-by: Jacky Lee <jacky.lee@egistec.com>

dcpleung commented:
Usually this type of interrupts is to signal when the TX FIFO _becomes_ empty so you can refill it with more data for long transmissions. So the usual workflow is to have TX FIFO empty interrupt disabled when not actively transmitting.

jackylee-go commented:
> Usually this type of interrupts is to signal when the TX FIFO _becomes_ empty so you can refill it with more data for long transmissions. So the usual workflow is to have TX FIFO empty interrupt disabled when not actively transmitting.

Of course, it's true.
But you can refer to the following code in "shell_uart.c", [#196](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/shell/backends/shell_uart.c#L196) [#L375](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/shell/backends/shell_uart.c#L375) 
This expects that when calling "tx_irq_enbale", if "FIFO" is empty, an interrupt callback should be received.
```
// 196
static void uart_callback(const struct device *dev, void *user_data)
{
	struct shell_uart_int_driven *sh_uart = (struct shell_uart_int_driven *)user_data;

	uart_irq_update(dev);

	if (uart_irq_rx_ready(dev)) {
		uart_rx_handle(dev, sh_uart);
	}

	if (uart_irq_tx_ready(dev)) {
		uart_tx_handle(dev, sh_uart);
	}
}

// 375
static int irq_write(struct shell_uart_int_driven *sh_uart,
		 const void *data, size_t length, size_t *cnt)
{
	*cnt = ring_buf_put(&sh_uart->tx_ringbuf, data, length);

	if (atomic_set(&sh_uart->tx_busy, 1) == 0) {
		uart_irq_tx_enable(sh_uart->common.dev);
	}

	return 0;
}
```
It first puts the data into `tx_ringbuf` and enables tx irq, then expects uart_callback to be called to execute `uart_tx_handle` to move data to `uart_fifo_fill` form `tx_ringbuf`.

There are similar solutions can be found on other serial drivers, with the same reasons.\
[uart_nrfx_uart.c#L845](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/serial/uart_nrfx_uart.c#L845)\
[uart_pl011.c#L353](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/serial/uart_pl011.c#L353)

Commit:
drivers: serial: ns16550: Fix TX IRQ not triggered when FIFO is empty

When uart_ns16550_irq_tx_enable() is called and the TX FIFO is already
empty, no new interrupt is generated, causing data transmission to stall
in some cases. This patch introduces a workaround to simulate an ISR
callback if the FIFO is empty when enabling the TX IRQ.

Signed-off-by: Jacky Lee <jacky.lee@egistec.com>

dcpleung commented:
Could you make it into a kconfig? So this can be enabled if needed, regardless of architecture.

jackylee-go commented:
> Could you make it into a kconfig? So this can be enabled if needed, regardless of architecture.

Do you mean to change it to something like "UART_NS16550_WA_ISR_REENABLE_INTERRUPT" in ["Kconfig.ns16550"](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/serial/Kconfig.ns16550#L90) ?.
```
menu "NS16550 Workarounds"

config UART_NS16550_WA_ISR_REENABLE_INTERRUPT
	bool "Re-enable interrupts by toggling IER at end of ISR"
	depends on UART_INTERRUPT_DRIVEN
	help
	  In some configurations (e.g. edge interrupt triggers), ...
```

jackylee-go commented:
> or something like `CONFIG_NS16550_WA_TX_FIFO_EMPTY_INTERRUPT`. This is not a architecture specific driver. So we want to avoid having architecture specific code in it unless absolutely necessary.

Understand.

But I believe that setting a `pending interrupt` is the ideal solution for this limitation.
However, manually setting a `pending interrupt` is architecture-dependent. For example, in [uart_nrfx_uart.c](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/serial/uart_nrfx_uart.c#L845), it uses `NVIC_SetPendingIRQ(IRQN)`.

Thus for platforms other than AE350, we’ve adopted an approach similar to [uart_pl011.c](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/serial/uart_pl011.c#L379), which directly invokes the `isr callback`.
If we don’t differentiate by platform here, then the only feasible solution would be to directly call the `isr callback`.

Although I’ve verified that the `isr callback` approach works correctly.
But, I still believe that setting a `pending interrupt` is the more appropriate solution.
That's because of it involves considerations such as ensuring that the callback runs in the `ISR` thread (e.g., stack size).

Or, could I let the user choose which solution to use via `Kconfig`?

dcpleung commented:
I would rather not have architecture specific code in the UART driver to only support one or two platforms. It will become a maintenance nightmare.

dcpleung commented on drivers/serial/Kconfig.ns16550:
```diff
@@ -103,6 +103,17 @@ config UART_NS16550_WA_ISR_REENABLE_INTERRUPT
 
+config UART_NS16550_WA_TX_FIFO_EMPTY_INTERRUPT
+	bool "Callback directly when the TX FIFO is already empty"
+	default y
```
This should not default to be enabled, and should only be enabled when needed.

dcpleung replied:
You can add a condition to enable by default if the shell UART backend is enabled.
